gio/src/cancellable.rs

```diff
@@ -1,10 +1,238 @@
+use futures_core::Future;
+use glib::{translate::*, IsA, Source};
+use std::num::NonZeroU64;
 
-#[test]
-fn check_callback() {
-    let c = crate::Cancellable::new();
-    c.connect_cancelled(|_| {});
-    c.cancel(); // if it doesn't crash at this point, then we're good to go!
+impl Cancellable {
+    // rustdoc-stripper-ignore-next
+    /// Creates a source that will never trigger.
+    pub fn pending_source() -> Source {
+        unsafe { from_glib_full(ffi::g_cancellable_source_new(std::ptr::null_mut())) }
+    }
+}
+
+// rustdoc-stripper-ignore-next
+/// The id of a cancelled handler that is returned by `CancellableExtManual::connect`. This type is
+/// analogous to [`glib::SignalHandlerId`].
+#[derive(Debug, Eq, PartialEq)]
+#[repr(transparent)]
+pub struct CancelledHandlerId(NonZeroU64);
+
+impl CancelledHandlerId {
+    // rustdoc-stripper-ignore-next
+    /// Returns the internal signal handler ID.
+    pub unsafe fn as_raw(&self) -> libc::c_ulong {
+        self.0.get() as libc::c_ulong
+    }
+}
+
+impl TryFromGlib<libc::c_ulong> for CancelledHandlerId {
+    type Error = GlibNoneError;
+    #[inline]
+    unsafe fn try_from_glib(val: libc::c_ulong) -> Result<Self, GlibNoneError> {
+        NonZeroU64::new(val as u64).map(Self).ok_or(GlibNoneError)
+    }
+}
+
+pub trait CancellableExtManual {
+    // rustdoc-stripper-ignore-next
+    /// Convenience function to connect to the `signal::Cancellable::cancelled` signal. Also
+    /// handles the race condition that may happen if the cancellable is cancelled right before
+    /// connecting. If the operation is cancelled from another thread, `callback` will be called
+    /// in the thread that cancelled the operation, not the thread that is running the operation.
+    /// This may be the main thread, so the callback should not do something that can block.
+    ///
+    /// `callback` is called at most once, either directly at the time of the connect if `self` is
+    /// already cancelled, or when `self` is cancelled in some thread.
+    ///
+    /// Since GLib 2.40, the lock protecting `self` is not held when `callback` is invoked. This
+    /// lifts a restriction in place for earlier GLib versions which now makes it easier to write
+    /// cleanup code that unconditionally invokes e.g.
+    /// [`CancellableExt::cancel()`][crate::prelude::CancellableExt::cancel()].
+    ///
+    /// # Returns
+    ///
+    /// The id of the signal handler or `None` if `self` has already been cancelled.
+    #[doc(alias = "g_cancellable_connect")]
+    fn connect_cancelled<F: FnOnce(&Self) + Send + Sync + 'static>(
+        &self,
+        callback: F,
+    ) -> Option<CancelledHandlerId>;
+    // rustdoc-stripper-ignore-next
+    /// Same as [`Self::connect_cancelled`] but takes a non-`Send+Sync` closure. If the cancellable
+    /// is cancelled from a different thread than it was connected to then the cancellation will
+    /// panic.
+    fn connect_cancelled_local<F: FnOnce(&Self) + 'static>(
```

Comment:
Did you find a use case for this?

Reply:
I just got rid of it, i can be added back if anyone asks for it

Reply:
Makes sense. Let's not add something that doesn't seem very useful :)